Support Untracked changes only status for the Commit Icon.

Drake103 opened the pull request:
Fixes #3887.

Changes proposed in this pull request:
 - Add new Icon for the `UntrackedChangesOnly` status
 - Extract icon selection logic to `CommitButtonStatusProvider`
 - Introduce `ICommitButtonStatus` and a separate class for each status.
 
Screenshots before and after (if PR changes UI):
before:
![image](https://user-images.githubusercontent.com/1637680/31413704-541be50e-ae23-11e7-8f55-f5245f18e946.png)
after (pale green):
![image](https://user-images.githubusercontent.com/1637680/31413698-4c685086-ae23-11e7-9c9c-3a0ab6a7bbd4.png)

How did I test this code:
 - Cover `CommitButtonStatusProvider` with unit tests
 - Manual testing

Has been tested on (remove any that don't apply):
 - GIT 2.10 and above
 - Windows 10

RussKie commented:
I feel that with the last commit you over-complicated the design.

Each status is just that - a single piece of information represented by an image. 
From the point of interface driven development DTOs/POCOs do not need interfaces because they have no behaviors. They may warrant an interface for abstraction reasons, but not in this case.

So a change which spanned 4 files now spans across 15 (slightly less if we move interfaces to implementations but still)...

I like your implementation for `CommitIconProvider` - I think it is sufficient.

Drake103 commented:
Main reason: I will have to test `CommitIconProvider` for concete Bitmaps (I don't like testing for concrete Resources).
Each status represents a Commit button apperance. My point is that Commit button states are presented not only with Image but also with Labels (Commit vs Commit (5)). Different images are provided using dynamic dispatch, this seems a behavior for me.
I agree, that providing only `Image` is really over-complication. Should I revert it?

RussKie commented:
> I will have to test CommitIconProvider for concete Bitmaps

I have no problems with that, these bitmaps are already compiled into the assembly anyway.

> Different images are provided using dynamic dispatch, this seems a behavior for me.

That's the behavior of the dispatcher not the images per se.

You could probably change the `CommitIconProvider` to `CommitStatusProvider` which returns `CommitStatus` poco object that could contain the status image and status format.
Not sure how this useful this may be, up to you if you wish to explore it further.

I also do not see any reason to wrap the images into properties - `Properties.Resources.xxx` are already static, extra layer just adds extra CPU cycles. 

```
 +        private static Bitmap IconClean
 +        {
 +            get { return Properties.Resources.IconClean; }
 +        }
```

> I agree, that providing only Image is really over-complication. Should I revert it?

Yes, less code the better - less code to maintain, less bugs to fix :)

Drake103 commented:
I have dropped the last commit and changed tests a bit to check for Bitmaps. Unfortunately each Bitmap resource `get` creates new instance of `Bitmap` class. Default `Equals` returns `false`, since they are reference types. I decided to share Bitmaps using public static readonly fields to avoid writing custom equaler.

Drake103 commented:
Fixed small glich with CommitButton:
before:
![image(https://user-images.githubusercontent.com/1637680/31413704-541be50e-ae23-11e7-8f55-f5245f18e946.png)
after:
![image](https://user-images.githubusercontent.com/1637680/31575540-f33de53a-b0f1-11e7-95ec-efc1b65a9501.png)

RussKie commented:
> Unfortunately each Bitmap resource get creates new instance of Bitmap class. 

I'm learning something new each day (or remembering things I've long forgotten) - Winforms image management is shocking...
https://theartofdev.com/2013/06/03/optimizing-images-usage-in-net-part-2/

I guess it is an area for us to improve in the future.

RussKie commented on GitUI/UserControls/ToolStripClasses/CommitIconProvider.cs:
```diff
@@ -0,0 +1,63 @@
+
+        public static readonly Bitmap IconMixed = Properties.Resources.IconMixed;
+
+        public static readonly Bitmap IconUntrackedOnly = Properties.Resources.IconUntrackedOnly;
```
these are probably best made internal

jbialobr commented:
Why did you choose green color? For me, green should be reserved for 'no changes, all clear and ok status'.

Drake103 commented:
I used pale green, because for me it was like "no tracked changes/almost no changes" (e.g. can checkout without resetting local directory). Another color I consider is violet, since it will be unique. Tell me the color, I will change it.

jbialobr commented:
Choose any unique color you like.

jbialobr commented:
Thank you.
